coordinator/src/test_coordinator.rs

```diff
@@ -0,0 +1,96 @@
+use super::context::SubStorageAccess;
+use super::validator::*;
+use ctypes::{CompactValidatorSet, ConsensusParams};
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+// Coordinator dedicated for mempool and miner testing
+// Only check_transactions, remove_transactions,
+pub struct TestCoordinator {
+    validator_set: CompactValidatorSet,
+    consensus_params: ConsensusParams,
+    body_count: AtomicUsize,
+    body_size: AtomicUsize,
+}
+
+impl Default for TestCoordinator {
+    fn default() -> Self {
+        Self {
+            validator_set: Default::default(),
+            consensus_params: ConsensusParams::default_for_test(),
+            body_count: AtomicUsize::new(0),
+            body_size: AtomicUsize::new(0),
+        }
+    }
+}
+
+impl Validator for TestCoordinator {
+    fn initialize_chain(&self, app_state: String) -> (CompactValidatorSet, ConsensusParams) {
+        unimplemented!()
+    }
+    fn open_block(&self, _context: &mut dyn SubStorageAccess, _header: &Header, _verified_crime: &[VerifiedCrime]) {
+        self.body_count.store(0, Ordering::SeqCst);
+        self.body_size.store(0, Ordering::SeqCst);
+    }
+
+    fn execute_transactions(&self, _context: &mut dyn SubStorageAccess, transactions: &[Transaction]) {
+        self.body_count.fetch_add(transactions.len(), Ordering::SeqCst);
+        let body_size: usize = transactions.iter().map(|tx| tx.size()).sum();
+        self.body_size.fetch_add(body_size, Ordering::SeqCst);
+    }
+
+    fn close_block(&self, context: &mut dyn SubStorageAccess) -> BlockOutcome {
+        let is_success = self.body_size.load(Ordering::SeqCst) > self.consensus_params.max_body_size();
+        BlockOutcome {
+            is_success,
+            updated_validator_set: self.validator_set.clone(),
+            updated_consensus_params: self.consensus_params,
+            transaction_results: (0..self.body_count.load(Ordering::SeqCst))
+                .map(|_| TransactionExecutionOutcome {
+                    events: Vec::new(),
+                })
+                .collect(),
+            events: Vec::new(),
+        }
+    }
+
+    fn check_transaction(&self, transaction: &Transaction) -> Result<(), ErrorCode> {
+        if transaction.size() > self.consensus_params.max_body_size() {
+            Err(1)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn fetch_transactions_for_block<'a>(
+        &self,
+        transactions: Vec<&'a TransactionWithMetadata>,
```

Comment:
Because the item type of the vector is a reference, it doesn't need to be a vector. How about using a slice?

Reply:
I used vector type to move references instead of copying them. 
The input references will be moved to inside of `TransactionWithGas` type.
I think it's not that meaningful. I'll use the slice.

Reply:
`rust
&'a [&'a TransactionWithMetadata]
`
Moving the reference and copying the reference is the same thing.

Reply:
Okay. I'll fix it. Thanks.